### PR TITLE
Adding Retry for UpdateStream LRO operation failures

### PR DESCRIPTION
**Pull request by VardhanThigle**
# Adding Retry for datastream update failure (after LRO poll)

> It's a good idea to open an issue first for discussion.

- [X] Tests pass
   -  Have tested this PR against orchestration for 152 streams.
- [ ] Appropriate changes to README are included in PR
   -   N/A
- [ ] TODO:
   -  Unit tested the Retry Strategy.
   - Have also added accessor and tests for resource manager towards making it easier (in future) to refactor and UT the launch_stream code.
   - Requesting to allow this PR to get merged for easy handover of the load testing.

### CODE REVIEW

File: streaming/streaming.go
Comment (manitgupta) [inline streaming/streaming.go:191]: Why do we need a separate configuration for LRO exponential backoff retry and normal API retries?

**Review by manitgupta — COMMENTED** [review]
There is a JSON file thats part of the commit, I think that needs to be removed?

File: streaming/streaming.go
Comment (VardhanThigle) [reply streaming/streaming.go:191]: The configuration parameters for retries when an API call errors out could be quite different than the retry when an LRO errors out and we should not tightly couple them.
For example an API could be retried after an initial delay of say 300 milliseconds and ramped up to an amount with exponential backoff, an LRO poll typically takes a few minutes to complete, so the initial retry could be in minutes instead of milliseconds.

File: streaming/streaming.go
Comment (VardhanThigle) [reply streaming/streaming.go:617]: At the time of raising this PR, I was waiting for the initial accessor PR to be merged in (it was under review), since the 152 streams test got delayed a bit, I can do a more fuller integration with accessors now on this PR and will update the code in my next push.

File: streaming/streaming.go
Comment (VardhanThigle) [reply streaming/streaming.go:189]: We have most of the basic stuff needed to: 
1. Add accessor for LaunchStream
2. Rerfactor launchsetream and add ut.

I need to think through 2 things though:
1. Should we have a single accessor for laucnh stream (this involves both create and update calls)
2. Should we have 2 separate accessors
3. Should resourcemanager accessor be chained from outside or should the call to reasourcemanager accessor be made outside and only output passed to the datastream_accessor.

This is the place (making 2 API calls for 1 client and 1 API call for another client) where we probably desperately need a DI framework :(

File: streaming/streaming.go
Comment (VardhanThigle) [reply streaming/streaming.go:606]: We have most of the basic stuff needed to:

Add accessor for LaunchStream
Rerfactor launchsetream and add ut.

I need to think through 2 things though:

Should we have a single accessor for laucnh stream (this involves both create and update calls)
Should we have 2 separate accessors
Should resourcemanager accessor be chained from outside or should the call to reasourcemanager accessor be made outside and only output passed to the datastream_accessor.

This is the place (making 2 API calls for 1 client and 1 API call for another client) where we probably desperately need a DI framework :(

**Comment by VardhanThigle** [comment]
> There is a JSON file thats part of the commit, I think that needs to be removed?

Thanks a ton. That was an uncleaned side effect of the load test :).
I cleaned it up from the PR.

File: accessors/resourcemanager/resourcemanager_accessor.go
Comment (manitgupta) [inline accessors/resourcemanager/resourcemanager_accessor.go:38]: This is still not fixed on Datastream end? Can we remove this?

**Review by manitgupta — APPROVED** [review]

File: streaming/streaming.go
Comment (VardhanThigle) [reply streaming/streaming.go:606]: Resolving as discssed offline. Checking this in to unblock handover of load testing. UT refactoring would be done later on.

**Pull request merged**
